File: lib/bracketResolver.ts
import {
  Match,
  Team,
  PredictionMap,
  GroupStanding,
  MatchConductData,
  GROUP_LETTERS,
  calculateGroupStandings,
  resolveAllR32Matches,
  getKnockoutWinner,
  getKnockoutLoser,
  getBest8ThirdPlaceTeams,
  rankThirdPlaceTeams,
  type RoundKey,
  ROUND_MATCH_STAGES,
} from './tournament'
import { resolveEntryPodiumPick, type PredictedPodium, type PredictionMode } from './podium'

export type BracketResult = {
  allGroupStandings: Map<string, GroupStanding[]>
  knockoutTeamMap: Map<number, { home: GroupStanding | null; away: GroupStanding | null }>
  champion: GroupStanding | null
  runnerUp: GroupStanding | null
  thirdPlace: GroupStanding | null
  qualifiedTeamIds: Set<string>
}

function extractMatchNumber(placeholder: string | null): number | null {
  if (!placeholder) return null
  const match = placeholder.match(/(?:Match\s*)?(\d+)/i)
  return match ? parseInt(match[1]) : null
}

/**
 * Shared core: computes group standings for all 12 groups, resolves R32 via
 * Annex C, then cascades through R16, QF, SF, third-place, and final.
 *
 * INTERNAL — not exported. Callers must pick an intent-named wrapper:
 *   - resolvePredictedBracket (from predictions; conduct intentionally absent)
 *   - resolveActualBracket    (from real results; conduct is a real tiebreaker)
 * This is what keeps real-world card data from ever leaking into a predicted
 * table — the divergence behind the knockout tie-break scoring bug.
 */
function resolveBracketCore(params: {
  matches: Match[]
  predictionMap: PredictionMap
  teams: Team[]
  conductData?: MatchConductData[]
}): BracketResult {
  const { matches, predictionMap, teams, conductData } = params

  // 1. Calculate group standings for all 12 groups
  const allGroupStandings = new Map<string, GroupStanding[]>()
  for (const letter of GROUP_LETTERS) {
    const gMatches = matches.filter(m => m.stage === 'group' && m.group_letter === letter)
    allGroupStandings.set(
      letter,
      calculateGroupStandings(letter, gMatches, predictionMap, teams, conductData)
    )
  }

  // 2. Resolve R32 matches via Annex C
  const r32Resolutions = resolveAllR32Matches(allGroupStandings)

  // 3. Build knockout team map starting with R32
  const knockoutTeamMap = new Map<number, { home: GroupStanding | null; away: GroupStanding | null }>()
  for (const [matchNum, teamsResolved] of r32Resolutions) {
    knockoutTeamMap.set(matchNum, teamsResolved)
  }

  const getMatch = (num: number) => matches.find(m => m.match_number === num)

  // 4. Cascade through knockout stages
  const resolveStage = (stage: string, isLoser = false) => {
    const stageMatches = matches
      .filter(m => m.stage === stage)
      .sort((a, b) => a.match_number - b.match_number)

    for (const m of stageMatches) {
      const homeMatchNum = extractMatchNumber(m.home_team_placeholder)
      const awayMatchNum = extractMatchNumber(m.away_team_placeholder)
      const homeSource = homeMatchNum ? knockoutTeamMap.get(homeMatchNum) : null
      const awaySource = awayMatchNum ? knockoutTeamMap.get(awayMatchNum) : null
      const homeSourceMatch = homeMatchNum ? getMatch(homeMatchNum) : null
      const awaySourceMatch = awayMatchNum ? getMatch(awayMatchNum) : null

      const resolveFn = isLoser ? getKnockoutLoser : getKnockoutWinner

      const home = homeSourceMatch && homeSource
        ? resolveFn(homeSourceMatch.match_id, predictionMap, homeSource.home, homeSource.away)
        : null
      const away = awaySourceMatch && awaySource
        ? resolveFn(awaySourceMatch.match_id, predictionMap, awaySource.home, awaySource.away)
        : null

      knockoutTeamMap.set(m.match_number, { home, away })
    }
  }

  resolveStage('round_16')
  resolveStage('quarter_final')
  resolveStage('semi_final')
  resolveStage('third_place', true)  // third-place takes losers of semi-finals
  resolveStage('final')

  // 5. Determine champion, runner-up, third place
  const finalMatch = matches.find(m => m.stage === 'final')
  const thirdPlaceMatch = matches.find(m => m.stage === 'third_place')

  let champion: GroupStanding | null = null
  let runnerUp: GroupStanding | null = null
  let thirdPlaceWinner: GroupStanding | null = null

  if (finalMatch) {
    const finalTeams = knockoutTeamMap.get(finalMatch.match_number)
    if (finalTeams) {
      champion = getKnockoutWinner(finalMatch.match_id, predictionMap, finalTeams.home, finalTeams.away)
      runnerUp = getKnockoutLoser(finalMatch.match_id, predictionMap, finalTeams.home, finalTeams.away)
    }
  }

  if (thirdPlaceMatch) {
    const thirdTeams = knockoutTeamMap.get(thirdPlaceMatch.match_number)
    if (thirdTeams) {
      thirdPlaceWinner = getKnockoutWinner(thirdPlaceMatch.match_id, predictionMap, thirdTeams.home, thirdTeams.away)
    }
  }

  // 6. Compute the set of 32 qualified team IDs (top 2 per group + best 8 thirds)
  const qualifiedTeamIds = new Set<string>()
  for (const [, standings] of allGroupStandings) {
    if (standings.length >= 2) {
      qualifiedTeamIds.add(standings[0].team_id)
      qualifiedTeamIds.add(standings[1].team_id)
    }
  }
  const best8 = getBest8ThirdPlaceTeams(allGroupStandings)
  for (const t of best8) {
    qualifiedTeamIds.add(t.team_id)
  }

  return {
    allGroupStandings,
    knockoutTeamMap,
    champion,
    runnerUp,
    thirdPlace: thirdPlaceWinner,
    qualifiedTeamIds,
  }
}

/**
 * Resolve a bracket from a user's PREDICTIONS.
 *
 * A predicted table is a pure function of the predicted scorelines — it carries
 * no card data, so conduct is deliberately not accepted. A perfectly tied
 * predicted group falls through to FIFA ranking (our deterministic stand-in for
 * FIFA's drawing of lots). Use this everywhere a bracket is derived from
 * predictions (scoring, bonuses, display, shadow) so all paths agree.
 */
export function resolvePredictedBracket(params: {
  matches: Match[]
  predictionMap: PredictionMap
  teams: Team[]
}): BracketResult {
  return resolveBracketCore(params)
}

/**
 * Resolve the ACTUAL bracket from real match results.
 *
 * Real fair-play/conduct IS a FIFA group tiebreaker, so conductData is required
 * here (pass [] only when it is genuinely unavailable). Predictions are scored
 * against this bracket.
 */
export function resolveActualBracket(params: {
  matches: Match[]
  predictionMap: PredictionMap
  teams: Team[]
  conductData: MatchConductData[]
}): BracketResult {
  return resolveBracketCore(params)
}

/**
 * For progressive mode: resolve knockout matches using actual team assignments
 * from the matches table (set by advance-teams flow) rather than user predictions.
 * Returns a map of match_number → { home, away } with team info as GroupStanding-like objects.
 */
export function resolveMatchesFromActual(
  matches: Match[],
  teams: Team[],
  roundKey: RoundKey
): Map<number, { home: GroupStanding | null; away: GroupStanding | null }> {
  const stages = ROUND_MATCH_STAGES[roundKey] ?? []
  const stageMatches = matches.filter(m => stages.includes(m.stage))
  const teamMap = new Map(teams.map(t => [t.team_id, t]))

  const result = new Map<number, { home: GroupStanding | null; away: GroupStanding | null }>()

  for (const match of stageMatches) {
    const homeTeam = match.home_team_id ? teamMap.get(match.home_team_id) : null
    const awayTeam = match.away_team_id ? teamMap.get(match.away_team_id) : null

    result.set(match.match_number, {
      home: homeTeam ? teamToStanding(homeTeam) : null,
      away: awayTeam ? teamToStanding(awayTeam) : null,
    })
  }

  return result
}

/** Convert a Team to a GroupStanding-like object for use in knockout forms */
function teamToStanding(team: Team): GroupStanding {
  return {
    team_id: team.team_id,
    country_name: team.country_name,
    country_code: team.country_code,
    flag_url: team.flag_url,
    group_letter: team.group_letter,
    fifa_ranking_points: team.fifa_ranking_points,
    played: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goalsFor: 0,
    goalsAgainst: 0,
    goalDifference: 0,
    points: 0,
  }
}

/**
 * Build a PredictionMap from actual match results (completed matches).
 * This is used to run the bracket resolver on "what actually happened"
 * so we can compare predicted vs actual brackets for bonus calculation.
 */
export function buildActualResultsMap(matches: Match[]): PredictionMap {
  const map = new Map<string, { home: number | null; away: number | null; homePso?: number | null; awayPso?: number | null; winnerTeamId?: string | null }>()

  for (const m of matches) {
    if (m.stage === 'group') {
      // For group matches, only include completed ones
      if ((m as any).is_completed && (m as any).home_score_ft != null) {
        map.set(m.match_id, {
          home: (m as any).home_score_ft,
          away: (m as any).away_score_ft,
          homePso: null,
          awayPso: null,
          winnerTeamId: null,
        })
      }
    } else {
      // For knockout matches, include completed ones with PSO data
      if ((m as any).is_completed && (m as any).home_score_ft != null) {
        map.set(m.match_id, {
          home: (m as any).home_score_ft,
          away: (m as any).away_score_ft,
          homePso: (m as any).home_score_pso ?? null,
          awayPso: (m as any).away_score_pso ?? null,
          winnerTeamId: (m as any).winner_team_id ?? null,
        })
      }
    }
  }

  return map
}

export type { PredictedPodium } from './podium'

/**
 * One-call predicted podium for an entry — the SAME derivation the scoring
 * engine applies, so display and points can never disagree about what a member
 * picked.
 *
 * This is a thin adapter: it assembles the brackets and hands off to
 * `resolveEntryPodiumPick` in lib/podium.ts, which owns the per-mode rules.
 * `predictionMode` is required — a caller that does not know the mode cannot
 * compute a correct podium, and defaulting it is what caused the incident.
 */
export function computeEntryPredictedPodium(params: {
  matches: Match[]
  predictionMap: PredictionMap
  teams: Team[]
  conductData?: MatchConductData[]
  predictionMode: PredictionMode
}): PredictedPodium {
  const { matches, predictionMap, teams, conductData = [], predictionMode } = params
  if (predictionMode === 'bracket_picker') {
    return { champion: null, runnerUp: null, thirdPlace: null }
  }

  const predictedBracket = resolvePredictedBracket({ matches, predictionMap, teams })

  // Progressive members predict the REAL fixtures, so their picks must be read
  // against the actual knockout pairings rather than their own cascade.
  const actualKnockoutTeamMap =
    predictionMode === 'progressive'
      ? resolveActualBracket({
          matches,
          predictionMap: buildActualResultsMap(matches),
          teams,
          conductData,
        }).knockoutTeamMap
      : undefined

  return resolveEntryPodiumPick({
    mode: predictionMode,
    matches,
    predictionMap,
    predictedBracket,
    actualKnockoutTeamMap,
  })
}

File: ios/OfficePools/Utilities/BracketResolver.swift
import Foundation

// MARK: - Bracket Result

struct BracketResult {
    /// Group standings keyed by group letter (A-L)
    let allGroupStandings: [String: [GroupStanding]]

    /// Knockout team assignments keyed by match number.
    /// Each value contains the home and away team for that knockout match.
    let knockoutTeamMap: [Int: (home: GroupStanding?, away: GroupStanding?)]

    /// Tournament champion (winner of the final)
    let champion: GroupStanding?

    /// Tournament runner-up (loser of the final)
    let runnerUp: GroupStanding?

    /// Third-place winner (winner of the third-place match)
    let thirdPlace: GroupStanding?

    /// The 8 best third-place teams that qualified for R32
    let qualifiedThirds: [GroupStanding]

    /// All 12 third-place teams ranked
    let rankedThirds: [GroupStanding]
}

// MARK: - Bracket Resolver

/// Resolves the full tournament bracket from a set of predictions.
///
/// Computes group standings for all 12 groups, resolves R32 via Annex C,
/// then cascades through R16, QF, SF, third-place, and final to determine
/// champion, runner-up, third place, and all knockout team assignments.
func resolveFullBracket(
    matches: [Match],
    predictions: [String: PredictionInput],
    teams: [Team]
) -> BracketResult {

    // 1. Calculate group standings for all 12 groups
    var allGroupStandings: [String: [GroupStanding]] = [:]
    for letter in GROUP_LETTERS {
        allGroupStandings[letter] = calculateGroupStandings(
            groupLetter: letter,
            matches: matches,
            predictions: predictions,
            teams: teams
        )
    }

    // 2. Rank all third-place teams
    let rankedThirds = rankThirdPlaceTeams(allGroupStandings: allGroupStandings)

    // 3. Get best 8 third-place teams
    let best8 = getBest8ThirdPlaceTeams(rankedThirds: rankedThirds)

    // 4. Resolve R32 matches
    var knockoutTeamMap: [Int: (home: GroupStanding?, away: GroupStanding?)] = [:]
    resolveR32Matches(
        allGroupStandings: allGroupStandings,
        best8: best8,
        into: &knockoutTeamMap
    )

    // 5. Cascade through knockout stages using KNOCKOUT_BRACKET
    resolveKnockoutStages(
        matches: matches,
        predictions: predictions,
        knockoutTeamMap: &knockoutTeamMap
    )

    // 6. Determine champion, runner-up, third place
    let champion: GroupStanding?
    let runnerUp: GroupStanding?
    let thirdPlaceWinner: GroupStanding?

    // Final is match 104
    if let finalMatch = matches.first(where: { $0.stage == "final" }),
       let finalTeams = knockoutTeamMap[finalMatch.matchNumber] {
        champion = bracketGetKnockoutWinner(
            matchId: finalMatch.matchId,
            predictions: predictions,
            homeTeam: finalTeams.home,
            awayTeam: finalTeams.away
        )
        runnerUp = bracketGetKnockoutLoser(
            matchId: finalMatch.matchId,
            predictions: predictions,
            homeTeam: finalTeams.home,
            awayTeam: finalTeams.away
        )
    } else {
        champion = nil
        runnerUp = nil
    }

    // Third-place match is match 103
    if let thirdPlaceMatch = matches.first(where: { $0.stage == "third_place" }),
       let thirdTeams = knockoutTeamMap[thirdPlaceMatch.matchNumber] {
        thirdPlaceWinner = bracketGetKnockoutWinner(
            matchId: thirdPlaceMatch.matchId,
            predictions: predictions,
            homeTeam: thirdTeams.home,
            awayTeam: thirdTeams.away
        )
    } else {
        thirdPlaceWinner = nil
    }

    return BracketResult(
        allGroupStandings: allGroupStandings,
        knockoutTeamMap: knockoutTeamMap,
        champion: champion,
        runnerUp: runnerUp,
        thirdPlace: thirdPlaceWinner,
        qualifiedThirds: best8,
        rankedThirds: rankedThirds
    )
}

// MARK: - R32 Resolution

/// Resolves all Round of 32 matches using Annex C for third-place team assignment
/// and R32_MATCHUPS for group winners/runners-up.
private func resolveR32Matches(
    allGroupStandings: [String: [GroupStanding]],
    best8: [GroupStanding],
    into knockoutTeamMap: inout [Int: (home: GroupStanding?, away: GroupStanding?)]
) {
    let matchNumbers = R32_MATCHUPS.keys.sorted()

    // First resolve all non-third-place slots (group winners & runners-up are deterministic)
    for matchNum in matchNumbers {
        guard let mapping = R32_MATCHUPS[matchNum] else { continue }
        let home = resolveNonThirdSlot(slot: mapping.homeSlot, allGroupStandings: allGroupStandings)
        let away = resolveNonThirdSlot(slot: mapping.awaySlot, allGroupStandings: allGroupStandings)
        knockoutTeamMap[matchNum] = (home: home, away: away)
    }

    // Try Annex C deterministic assignment
    if best8.count == 8 {
        let qualifyingGroups = best8.map { $0.groupLetter }
        if let annexCResult = lookupAnnexC(qualifyingGroups: qualifyingGroups) {
            // Build map: group letter -> third-place team
            var thirdByGroup: [String: GroupStanding] = [:]
            for team in best8 {
                thirdByGroup[team.groupLetter] = team
            }

            // Apply Annex C assignments
            for (matchNum, thirdGroupLetter) in annexCResult {
                if let team = thirdByGroup[thirdGroupLetter],
                   let current = knockoutTeamMap[matchNum] {
                    // Third-place team is always the away team in these matches
                    knockoutTeamMap[matchNum] = (home: current.home, away: team)
                }
            }
            return
        }
    }

    // Fallback: backtracking (used when < 8 third-place teams or Annex C lookup fails)
    var thirdSlots: [(matchNum: Int, side: String, eligible: [String])] = []
    for matchNum in matchNumbers {
        guard let mapping = R32_MATCHUPS[matchNum] else { continue }
        if case .bestThird(let eligible) = mapping.homeSlot {
            thirdSlots.append((matchNum: matchNum, side: "home", eligible: eligible))
        }
        if case .bestThird(let eligible) = mapping.awaySlot {
            thirdSlots.append((matchNum: matchNum, side: "away", eligible: eligible))
        }
    }

    var assignment: [Int: GroupStanding] = [:]
    var usedTeamIds: Set<String> = []

    func backtrack(_ slotIdx: Int) -> Bool {
        if slotIdx == thirdSlots.count { return true }

        let slot = thirdSlots[slotIdx]
        for team in best8 {
            if usedTeamIds.contains(team.teamId) { continue }
            if !slot.eligible.contains(team.groupLetter) { continue }

            usedTeamIds.insert(team.teamId)
            assignment[slotIdx] = team

            if backtrack(slotIdx + 1) { return true }

            usedTeamIds.remove(team.teamId)
            assignment.removeValue(forKey: slotIdx)
        }
        return false
    }

    _ = backtrack(0)

    for i in 0..<thirdSlots.count {
        let slot = thirdSlots[i]
        let team = assignment[i]
        if let current = knockoutTeamMap[slot.matchNum] {
            if slot.side == "home" {
                knockoutTeamMap[slot.matchNum] = (home: team, away: current.away)
            } else {
                knockoutTeamMap[slot.matchNum] = (home: current.home, away: team)
            }
        }
    }
}

/// Resolves a non-third-place slot (group winner or group runner-up).
private func resolveNonThirdSlot(
    slot: SlotType,
    allGroupStandings: [String: [GroupStanding]]
) -> GroupStanding? {
    switch slot {
    case .groupWinner(let group):
        return allGroupStandings[group]?.first
    case .groupRunnerUp(let group):
        let standings = allGroupStandings[group]
        return (standings?.count ?? 0) >= 2 ? standings?[1] : nil
    case .bestThird:
        return nil // Handled separately
    }
}

// MARK: - Knockout Stage Cascading

/// Cascades through knockout stages (R16, QF, SF, third-place, final)
/// by reading home_team_placeholder / away_team_placeholder from match data
/// to determine source matches — matching the web app's approach.
private func resolveKnockoutStages(
    matches: [Match],
    predictions: [String: PredictionInput],
    knockoutTeamMap: inout [Int: (home: GroupStanding?, away: GroupStanding?)]
) {
    let stages = ["round_16", "quarter_final", "semi_final", "third_place", "final"]

    for stage in stages {
        let isLoserStage = (stage == "third_place")
        let stageMatches = matches
            .filter { $0.stage == stage }
            .sorted { $0.matchNumber < $1.matchNumber }

        for match in stageMatches {
            let homeMatchNum = extractMatchNumber(placeholder: match.homeTeamPlaceholder)
            let awayMatchNum = extractMatchNumber(placeholder: match.awayTeamPlaceholder)

            let homeSource = homeMatchNum.flatMap { knockoutTeamMap[$0] }
            let awaySource = awayMatchNum.flatMap { knockoutTeamMap[$0] }

            let homeSourceMatch = homeMatchNum.flatMap { num in matches.first(where: { $0.matchNumber == num }) }
            let awaySourceMatch = awayMatchNum.flatMap { num in matches.first(where: { $0.matchNumber == num }) }

            let resolveFn = isLoserStage ? bracketGetKnockoutLoser : bracketGetKnockoutWinner

            let home: GroupStanding? = if let sourceMatch = homeSourceMatch, let source = homeSource {
                resolveFn(sourceMatch.matchId, predictions, source.home, source.away)
            } else {
                nil
            }

            let away: GroupStanding? = if let sourceMatch = awaySourceMatch, let source = awaySource {
                resolveFn(sourceMatch.matchId, predictions, source.home, source.away)
            } else {
                nil
            }

            knockoutTeamMap[match.matchNumber] = (home: home, away: away)
        }
    }
}

// MARK: - Knockout Winner/Loser (GroupStanding-based)

/// Determines the winner of a knockout match based on predictions, returning a GroupStanding.
/// Falls back to FIFA ranking if the prediction results in a draw without PSO resolution.
private func bracketGetKnockoutWinner(
    matchId: String,
    predictions: [String: PredictionInput],
    homeTeam: GroupStanding?,
    awayTeam: GroupStanding?
) -> GroupStanding? {
    guard let home = homeTeam, let away = awayTeam else { return nil }
    guard let pred = predictions[matchId],
          let homeScore = pred.homeScore,
          let awayScore = pred.awayScore else { return nil }

    // Full-time winner
    if homeScore > awayScore { return home }
    if awayScore > homeScore { return away }

    // Tied: check PSO exact scores
    if let homePso = pred.homePso, let awayPso = pred.awayPso {
        if homePso > awayPso { return home }
        if awayPso > homePso { return away }
    }

    // Tied: check explicit winner team ID
    if let winnerTeamId = pred.winnerTeamId {
        if winnerTeamId == home.teamId { return home }
        if winnerTeamId == away.teamId { return away }
    }

    // Fallback: FIFA ranking
    return home.fifaRankingPoints >= away.fifaRankingPoints ? home : away
}

/// Determines the loser of a knockout match based on predictions, returning a GroupStanding.
private func bracketGetKnockoutLoser(
    matchId: String,
    predictions: [String: PredictionInput],
    homeTeam: GroupStanding?,
    awayTeam: GroupStanding?
) -> GroupStanding? {
    guard let home = homeTeam, let away = awayTeam else { return nil }
    guard let winner = bracketGetKnockoutWinner(
        matchId: matchId,
        predictions: predictions,
        homeTeam: home,
        awayTeam: away
    ) else { return nil }
    return winner.teamId == home.teamId ? away : home
}

// MARK: - Helper

/// Extracts a match number from a placeholder string like "Winner Match 73" or "Match 73".
func extractMatchNumber(placeholder: String?) -> Int? {
    guard let placeholder = placeholder else { return nil }
    let pattern = #"(?:Match\s*)?(\d+)"#
    guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
          let match = regex.firstMatch(
              in: placeholder,
              range: NSRange(placeholder.startIndex..., in: placeholder)
          ),
          let range = Range(match.range(at: 1), in: placeholder) else {
        return nil
    }
    return Int(placeholder[range])
}
